### PR TITLE
[WONTMERGE] Refactor how links / URLs work

### PR DESCRIPTION
**Pull request by jamesgorrie**
WONTMERGE - this was a POC, going to refactor the PR out.

(and restructure the way we do parsers / URLs / params / zombies / :trollface: )

## Step 1:
Use worksParams everywhere

## Step 2
Remove worksParams from everywhere except where it's needed. We've completely complicated things by passing the state around in the clientside URL and ahve to keep passing it about from page to page.

We can just keep the state in the worksUrl as that's where we need it.

It currently behaves a bit weirdly i.e. forgets page etc, so this should probably iron out a few of those things.

### CODE REVIEW

File: common/services/catalogue/codecs.js [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+// We extend from ApiSearchParams as this will be more robust for reporting
+// as we open up the `workType`s.
+
+// This way we are tracking exactly what people are seeing.
+export type TrackWorksParams = {|
+  ...ApiWorksParams,
+  source: ?string,
+|};
+
+function stringTocsv(s: ?string): ?(string[]) {
```
Comment (davidpmccormick) [inline common/services/catalogue/codecs.js]: Capital 'C' in `stringToCsv`?

File: common/services/catalogue/codecs.js [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+  }
+}
+
+// This is the object you get from ctx.query, at least a `{}`.
+type RequestQuery = { [key: string]: string };
+
+type NextRoute<T> = {|
+  fromQuery: (q: RequestQuery) => T,
+  link: (t: $Shape<T>) => NextLinkType,
+  query: (t: $Shape<T>) => RequestQuery,
```
Comment (davidpmccormick) [inline common/services/catalogue/codecs.js]:
```suggestion
  toQuery: (t: $Shape<T>) => RequestQuery,
```

File: common/services/catalogue/codecs.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        query: ItemCodec.query(as),
+      },
+    };
+  },
+  query(params) {
+    return serialiseUrl(params);
+  },
+};
+
+export const defaultWorkTypes = ['a', 'k', 'q', 'i', 'g'];
```
Comment (gestchild) [inline common/services/catalogue/codecs.js]:
```suggestion
export const defaultWorkTypes = ['a', 'g', 'i', 'k', 'l', 'q'];
```
we've added ephemera (l) to this

**Comment by jamesgorrie** [comment]
Close in favour of smaller PRs

**Pull request closed without merging**